commands/src/main/java/discord4j/commands/CommandBootstrapper.java

```diff
@@ -0,0 +1,168 @@
+package discord4j.commands;
+
+import discord4j.core.DiscordClient;
+import discord4j.core.event.domain.message.MessageCreateEvent;
+import reactor.core.publisher.Flux;
+import reactor.util.Logger;
+import reactor.util.Loggers;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Set;
+import java.util.concurrent.CopyOnWriteArraySet;
+
+/**
+ * The (thread-safe) entry point for actually using commands. To use, simply instantiate and call
+ * {@link #attach(discord4j.core.DiscordClient)}.
+ *
+ * To register events, obtain the {@link discord4j.commands.CommandDispatcher} via {@link #getCommandDispatcher()}
+ * or create a new {@link discord4j.commands.CommandDispatcher} instance replace the default instance with the new
+ * one via {@link #setCommandDispatcher(CommandDispatcher)}.
+ *
+ * @see #attach(discord4j.core.DiscordClient)
+ * @see discord4j.commands.CommandProvider
+ * @see discord4j.commands.CommandDispatcher
+ * @see discord4j.commands.CommandErrorHandler
+ * @see discord4j.commands.DefaultCommandDispatcher
+ */
+public final class CommandBootstrapper {
+
+    private final static Logger log = Loggers.getLogger(CommandBootstrapper.class);
+    private final static CommandDispatcher DEFAULT_DISPATCHER = new DefaultCommandDispatcher();
+    private final static CommandErrorHandler DEFAULT_ERROR_HANDLER = (context, error) -> {
+        log.warn("Command excecution failed! Reason: {}", error.response().orElse("None"));
+    };
+
+    private volatile CommandDispatcher dispatcher = DEFAULT_DISPATCHER;
+    private volatile CommandErrorHandler errorHandler = DEFAULT_ERROR_HANDLER;
+
+    private final Set<CommandProvider> providers;
+
+    /**
+     * Constructs the message listener for commands.
+     *
+     * @see #attach(discord4j.core.DiscordClient)
+     * @param providers The {@link discord4j.commands.CommandProvider}s to initialize this instance with. Note that
+     * this set is expected to be thread-safe as there will be concurrent read access required.
+     */
+    public CommandBootstrapper(Set<CommandProvider> providers) {
```

Comment:
Is it expected to mutate dispatcher and errorHandler after it's been set? If not, why not supply them here instead?

If you want to enforce thread safety, then copy them into its own structure, owned by this object. Or is it not that important?